src/ImageSharp/Formats/Png/Zlib/DeflaterEngine.cs

```diff
@@ -0,0 +1,853 @@
+
+                this.isDisposed = true;
+            }
+
+            GC.SuppressFinalize(this);
+        }
+
+        [MethodImpl(InliningOptions.ShortMethod)]
+        private void UpdateHash()
+        {
+            byte* pinned = this.pinnedWindowPointer;
+            this.insertHashIndex = (pinned[this.strstart] << DeflaterConstants.HASH_SHIFT) ^ pinned[this.strstart + 1];
+        }
+
+        /// <summary>
+        /// Inserts the current string in the head hash and returns the previous
+        /// value for this hash.
+        /// </summary>
+        /// <returns>The previous hash value</returns>
+        [MethodImpl(InliningOptions.ShortMethod)]
+        private int InsertString()
+        {
+            short match;
+            int hash = ((this.insertHashIndex << DeflaterConstants.HASH_SHIFT) ^ this.pinnedWindowPointer[this.strstart + (DeflaterConstants.MIN_MATCH - 1)]) & DeflaterConstants.HASH_MASK;
+
+            short* pinnedHead = this.pinnedHeadPointer;
+            this.pinnedPrevPointer[this.strstart & DeflaterConstants.WMASK] = match = pinnedHead[hash];
+            pinnedHead[hash] = unchecked((short)this.strstart);
+            this.insertHashIndex = hash;
+            return match & 0xFFFF;
+        }
+
+        private void SlideWindow()
+        {
+            Unsafe.CopyBlockUnaligned(ref this.window[0], ref this.window[DeflaterConstants.WSIZE], DeflaterConstants.WSIZE);
+            this.matchStart -= DeflaterConstants.WSIZE;
+            this.strstart -= DeflaterConstants.WSIZE;
+            this.blockStart -= DeflaterConstants.WSIZE;
+
+            // Slide the hash table (could be avoided with 32 bit values
+            // at the expense of memory usage).
+            short* pinnedHead = this.pinnedHeadPointer;
+            for (int i = 0; i < DeflaterConstants.HASH_SIZE; ++i)
+            {
+                int m = pinnedHead[i] & 0xFFFF;
+                pinnedHead[i] = (short)(m >= DeflaterConstants.WSIZE ? (m - DeflaterConstants.WSIZE) : 0);
+            }
+
+            // Slide the prev table.
+            short* pinnedPrev = this.pinnedPrevPointer;
+            for (int i = 0; i < DeflaterConstants.WSIZE; i++)
+            {
+                int m = pinnedPrev[i] & 0xFFFF;
+                pinnedPrev[i] = (short)(m >= DeflaterConstants.WSIZE ? (m - DeflaterConstants.WSIZE) : 0);
+            }
+        }
+
+        /// <summary>
+        /// <para>
+        /// Find the best (longest) string in the window matching the
+        /// string starting at strstart.
+        /// </para>
+        /// <para>
+        /// Preconditions:
+        /// <code>
+        /// strstart + DeflaterConstants.MAX_MATCH &lt;= window.length.</code>
+        /// </para>
+        /// </summary>
+        /// <param name="curMatch">The current match.</param>
+        /// <returns>True if a match greater than the minimum length is found</returns>
+        private bool FindLongestMatch(int curMatch)
```

Comment:
This appears to be a hotspot according to dotTrace. 

![image](https://user-images.githubusercontent.com/385879/68920030-1e370f80-07c8-11ea-9fb5-89b8e94c4992.png)